File: services/api/app/models/memory.py
import uuid
from datetime import datetime, timezone
from sqlalchemy import String, Integer, ForeignKey, Text, LargeBinary, DateTime
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import ARRAY
from ..core.database import Base


class Memory(Base):
    __tablename__ = "memories"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id: Mapped[str] = mapped_column(String, ForeignKey("users.id", ondelete="CASCADE"))
    project_id: Mapped[str | None] = mapped_column(String, ForeignKey("projects.id"), nullable=True)
    level: Mapped[str] = mapped_column(String, default="user")
    type: Mapped[str] = mapped_column(String, default="note")
    title: Mapped[str] = mapped_column(String)
    content: Mapped[str] = mapped_column(Text)
    embedding: Mapped[bytes | None] = mapped_column(LargeBinary, nullable=True)
    tags: Mapped[list] = mapped_column(ARRAY(String), default=list)
    source: Mapped[str | None] = mapped_column(String, nullable=True)
    importance: Mapped[int] = mapped_column(Integer, default=1)
    external_source: Mapped[str | None] = mapped_column(String, nullable=True)
    external_path: Mapped[str | None] = mapped_column(String, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    user = relationship("User")
    project = relationship("Project", back_populates="memories")

File: packages/types/src/memory.ts
export type MemoryLevel = 'session' | 'user' | 'project' | 'external'

export interface Memory {
  id: string
  userId: string
  projectId?: string
  level: MemoryLevel
  type: 'note' | 'decision' | 'fact' | 'preference' | 'learning' | 'task' | 'idea' | 'summary'
  title: string
  content: string
  embedding: number[]
  tags: string[]
  source?: string
  sourceUrl?: string
  importance: number
  externalSource?: 'obsidian' | 'google_drive' | 'notion'
  externalPath?: string
  createdAt: string
  updatedAt: string
}

export interface MemorySearchResult {
  memory: Memory
  relevance: number
  context?: string
}

export interface ObsidianVault {
  id: string
  name: string
  path: string
  notes: number
  lastSync: string
  connected: boolean
}
